src/primitives/bezier.jl

```diff
@@ -0,0 +1,103 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    BezierCurve(points)
+
+A recursive Bézier curve with control points `points`.
+See https://en.wikipedia.org/wiki/Bézier_curve.
+A point on the curve `b` can be evaluated by calling
+`b(t)` with `t` between `0` and `1`.
+The evaluation method defaults to DeCasteljau's algorithm
+for accurate evaluation. Horner's method, faster with a
+large number of points but less precise, can be used via
+`b(t, Horner())`.
+
+## Example
+
+```julia
+BezierCurve(Point2[(0.,0.),(1.,-1.)])
+```
+"""
+struct BezierCurve{Dim,T,V<:AbstractVector{Point{Dim,T}}} <: Primitive{Dim,T}
+  controls::V
+end
+
+BezierCurve(points::Vararg) = BezierCurve(collect(points))
+BezierCurve(points::AbstractVector{<:Tuple}) = BezierCurve(Point.(points))
+
+ncontrols(b::BezierCurve) = length(b.controls)
+
+degree(b::BezierCurve) = ncontrols(b) - 1
+
+"""
+Evaluation method used to obtain a point along
+a Bézier curve from a parametric expression.
+"""
+abstract type BezierEvalMethod end
+
+"""
+Accurate evaluation using De Casteljau's algorithm.
+Recommended for a small number of control points.
+
+See https://en.wikipedia.org/wiki/De_Casteljau%27s_algorithm.
+"""
+struct DeCasteljau <: BezierEvalMethod end
+
+"""
+Approximate evaluation using Horner's method.
+Recommended for a large number of control points,
+if you can afford a precision loss.
+
+See https://en.wikipedia.org/wiki/Horner%27s_method.
+"""
+struct Horner <: BezierEvalMethod end
+
+(curve::BezierCurve)(t) = curve(t, DeCasteljau())
+
+function (curve::BezierCurve)(t, ::DeCasteljau)
+  if t < 0 || t > 1
+    throw(DomainError(t, "b(t) is not defined for t outside [0, 1]."))
+  end
+  ss = collect(segments(Chain(curve.controls)))
+  points = [s(t) for s in ss]
+  if length(points) == 1
+    points[1]
+  else
+    BezierCurve(points)(t)
+  end
+end
+
+"""
+Apply Horner's method on the monomial representation of the
+Bézier curve B = ∑ᵢ aᵢtⁱ with i ∈ [0, n], n the degree of the
+curve, aᵢ = binomial(n, i) * pᵢ * t̄ⁿ⁻ⁱ and t̄ = (1 - t).
+Horner's rule recursively reconstructs B from a sequence bᵢ
+with bₙ = aₙ and bᵢ₋₁ = aᵢ₋₁ + bᵢ * t until b₀ = B.
+"""
+function (curve::BezierCurve{Dim,T})(t, ::Horner) where {Dim,T}
+  if t < 0 || t > 1
+    throw(DomainError(t, "b(t) is not defined for t outside [0, 1]."))
+  end
+  cs = curve.controls
+  t̄ = one(T) - t
+  n = degree(curve)
+  pₙ = @views coordinates(cs[end])
+  aₙ = pₙ
+
+  # initialization with i = n + 1
+  bᵢ₋₁ = bₙ = aₙ
+  cᵢ₋₁ = one(T)
+  t̄ⁿ⁻ⁱ = one(T)
+  for i in n:-1:1
+    cᵢ₋₁ *= i / (n - i + one(T))
+    pᵢ₋₁ = @views coordinates(cs[i])
+    t̄ⁿ⁻ⁱ *= t̄
+    aᵢ₋₁ = cᵢ₋₁ * pᵢ₋₁ * t̄ⁿ⁻ⁱ
+    bᵢ₋₁ = aᵢ₋₁ + bᵢ₋₁ * t
+  end
+
+  b₀ = bᵢ₋₁
```

Comment:
I would agree that this is useless, but it makes it clearer when comparing to the original algorithm.

Reply:
At least this one is being used later in the `Point` return. 👍🏽